### DISCUSSION TITLE
OpenSSL iOS Privacy Manifest

### DISCUSSION THREAD

**Discussion by dillonmasseyrm** [openssl/openssl#23262] [General Discussion]
I was wondering if there is going to be any support for privacy manifests with Apple requiring OpenSSL to provide one in the list stated here https://developer.apple.com/support/third-party-SDK-requirements/
https://developer.apple.com/documentation/bundleresources/privacy_manifest_files

**Comment by nate-at-king** [comment]
With [April 29 just around the corner](https://developer.apple.com/news/?id=fxu2qp7b), I was wondering the same as @dillonmasseyrm. Even a confirmation that this is the right post to follow to hear a yes/no/maybe in the future would be helpful.

**Comment by t8m** [comment]
I am sorry but we do not provide any binaries for iOS so we cannot provide these manifest files.

**Reply by kroeckx** [reply]
I'm wondering if we can at least provide some sample file that can be used by people making the binaries. For instance, we can at least list which functions we can call. The only one I see that we might potentially call is fstat(), were we might need to say it's reason `0A2A.1`.

**Reply by nate-at-king** [reply]
> I'm wondering if we can at least provide some sample file that can be used by people making the binaries. For instance, we can at least list which functions we can call. The only one I see that we might potentially call is fstat(), were we might need to say it's reason 0A2A.1.

That would be greatly appreciated.

**Reply by kroeckx** [reply]
We also have calls to stat()

Maybe we should actually request a different reason code, since it's not actually returning the timestamp. Someone should probably at least document which calls we have in the libraries and apps, and if they use the timestamp or not.

**Reply by jzilske** [reply]
FWIW, it’s not only about APIs being called (so-called required Reason APIs); that is only a subset of what is required (cf. eg. https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_data_use_in_privacy_manifests)

**Reply by tom-cosgrove-arm** [reply]
This really looks like something that people actually using OpenSSL in their iOS apps - who are presumably already doing this for their apps - could create a template for in a pull request, for OpenSSL folks to review and merge when agreed. Then other people using OpenSSL in their iOS apps could incorporate it into their binaries. i.e. this feels like something that should come from the appropriate part of the community

**Reply by kroeckx** [reply]
> FWIW, it’s not only about APIs being called (so-called required Reason APIs); that is only a subset of what is required (cf. eg. https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_data_use_in_privacy_manifests)

OpenSSL does not collect any of that data.

**Reply by nate-at-king** [reply]
> OpenSSL does not collect any of that data.

That is what makes it so helpful about having at least those reason codes documented somewhere. Sure it would be best for my situation (and those like me) if that "documentation" was a `PrivacyInfo.xcprivacy` file all ready to go. But I can work with whatever I get, this is all open source after all with all that entails for usage expectations.

Since i'm not genuinely expecting OpenSSL to ever collecting tracked data on its own ([as defined by apple](https://developer.apple.com/app-store/app-privacy-details/#user-tracking)) it means I mostly need a maintainable way of collecting those reason codes. That lets me avoid is creating a lengthy section in my team's internal "how to update OpenSSL" guide with details on how to take [this from Apple](https://developer.apple.com/documentation/bundleresources/privacy_manifest_files/describing_use_of_required_reason_api?language=objc) and compare that against the current release of OpenSSL to see if we need to add more reason codes to our own Privacy Manifest.

**Comment by wlxo0401** [comment]
I'm adding a static library of C and C++ to my iOS project by making it '.a'. 

The static library contains OpenSSL. Should I add 'Privacy Manifest' in that situation as well??

What should I do if OpenSSL doesn't add 'Privacy Manifest'?

**Reply by kroeckx** [reply]
My understanding is that you need to add a privacy manifest about openssl. You will need to create it yourself.

**Reply by wlxo0401** [reply]
> My understanding is that you need to add a privacy manifest about openssl. You will need to create it yourself.

@kroeckx 

Thank you for your answer. 

I know how to add 'Privacy Manifest' to my app. 

but 
1. I don't know how to add it to OpenSSL.
2. I don't know how to add it to the '.a' library.

My understanding of what Apple means is to add 'PrivacyInfo' to each library.

But my situation is 

OpenSSL -> Static Library(.a) -> My App

I think it's woven in order.

I also took over this project, so I can do an app, but I'm having a hard time because I don't make a library.

**Reply by nate-at-king** [reply]
Since you're linking OpenSSL as a `.a`, then to the best of my personal understanding in those situations you just need to make sure to add the information PrivacyManifest information for OpenSSL directly into your app's own Privacy Manifest file.

**Reply by wlxo0401** [reply]
> Since you're linking OpenSSL as a .a, then to the best of my personal understanding in those situations you just need to make sure to add the information PrivacyManifest information for OpenSSL directly into your app's own Privacy Manifest file.

@nate-at-king 

Yes, I am using OpenSSL in a library consisting of C and C++.

And I made it into '.a' and I'm using it in iOS project. 
(I'm not a library developer, it's a library that already existed. So I'm trying to 'Privacy Manifest' without knowing how to set up the library.)

I understood that 'PrivacyInfo' doesn't have to be added directly to OpenSSL or directly to static library, but I can add it to 'PrivacyInfo' to the project of the app myself.

I thought Apple announced it to mean that all libraries should contain 'PrivacyInfo' respectively, is this wrong???

I may not be able to understand your meaning because I still lack development knowledge and am not good at English. Nevertheless, thank you for your help.

**Comment by dillonmasseyrm** [comment]
Based on whats said, I'm assuming it will need to be self supplied. We use a fork of https://github.com/x2on/OpenSSL-for-iPhone , which is then included in our iOS app as an xcframework. So I will add a privacy manifest there, as that is what is including OpenSSL.

**Comment by jksmx** [comment]
I think openssl is a c static library . So if you build .a file to add to you project . You only need add a codesign to .a file . Do not use privacyInfo file . Because openssl does not use any ios api . If nesscessry add it , you can make use .a file to make a framework type .And add a template privacyInfo file to you new framework root path .

**Reply by wlxo0401** [reply]
There's a lot of disagreement about this part. '.a' is included in the app, so some say to add 'PrivacyManifest' to the app's 'PrivacyManifest'.

**Comment by wwforever-nw** [comment]
I understand that app developers do not need to build a new SDK or apply the built SDK due to the SDK signature in Apple's privacy manifest.
Currently, I use SDKs such as Firebase, Unity, Google Signin, and Appsflyer, but most SDKs do not provide updates but provide manifest guides.
It would be nice if openssl also provided guidance on issues related to Apple's privacy manifest.
Below is the appsflyer guide link.
https://support.appsflyer.com/hc/en-us/articles/21677433322641-Privacy-Manifest#required-reasons-api

**Comment by naahmedea** [comment]
Can someone please confirm if the below reason should be declared as part of OpenSSL privacy manifest file?

1. `0A2A.1` Declare this reason if your third-party SDK is providing a wrapper function around file timestamp API(s) for the app to use, and you only access the file timestamp APIs when the app calls your wrapper function. This reason may only be declared by third-party SDKs. This reason may not be declared if your third-party SDK was created primarily to wrap required reason API(s).

Information accessed for this reason, or any derived information, may not be used for your third-party SDK’s own purposes or sent off-device by your third-party SDK.

**Reply by t8m** [reply]
Although OpenSSL calls stat()/fstat() it does not use the timestamp data obtained via this call at all. The call is done for other purposes such as identifying whether a given file node exists and what type is it (i.e., is it a directory or regular file).

Thus I do not think this reason should be declared but I am not an iOS expert.

**Reply by naahmedea** [reply]
Since i could not find a privacy manifest file for OpenSSL, could you provide us with details that can be incorporated into a privacy manifest file for OpenSSL for our reference as you might have more context on OpenSSL's functionality?

**Comment by slontis** [comment]
Issue was raised here...
https://github.com/openssl/openssl/issues/23494

**Comment by slontis** [comment]
I dont see any reason why OpenSSL cant provide this file. See https://github.com/openssl/openssl/pull/24032
